### ISSUE TITLE
WCF - FaultException with FaultCode causes CommunicationException on client

### ISSUE THREAD

**Issue by dalefugier** [mono/mono#12995]
If a WCF server, running on Windows, throws a `FaultException` that contains a `FaultCode` to a WCF client, running on macOS, a `CommunicationException` is caught, not the expected `FaultException`.

Note, this is a **regression** an earlier version of Mono. I've tested against Version 4.0.2.

## Steps to Reproduce

You can find source to a simple example that reproduces his behavior here:

https://github.com/dalefugier/Echo

### Current Behavior

A `FaultException` that contains a `FaultCode`, thrown by a server, is caught as a  `CommunicationException` on a client.

### Expected Behavior

A `FaultException` that contains a `FaultCode`, thrown by a server, should be caught as a `FaultException` on a client.

## On which platforms did you notice this

[X] macOS
[ ] Linux
[ ] Windows

**Version Used**:

Mono Version 5.16.0.220

## Stacktrace

```
{System.ServiceModel.CommunicationException: Received an invalid SOAP Fault message ---
> System.Xml.XmlException: Content cannot be converted to the type
System.Xml.XmlQualifiedName. Line 1, position 78. ---> System.FormatException: The ':'
character, hexadecimal value 0x3A, cannot be included in a name.
```

**Comment by dalefugier** [comment]
If you'd rather not clone the repo, that contains the sample that replicted the issue, here is the solution:

[Echo.zip](https://github.com/mono/mono/files/2865723/Echo.zip)

Thanks,

-- Dale

**Comment by EgorBo** [comment]
Duplicates https://github.com/mono/mono/issues/10304

**Comment by dalefugier** [comment]
@EgorBo , @marek-safar, I'm hoping I can get come clarification on why the issue was marked as `Duplicate` and `Closed`. The linked issue ([10304)](https://github.com/mono/mono/issues/10304) is related to Android, not macOS. And I'm using a fairly new, stable version of Mono. Is it possible to get someone to  to test my `Echo` test harness with Xamarin.Mac 5.7 and confirm? If not, how can I get Xamarin.Mac 5.7?

Thanks,

-- Dale

**Comment by EgorBo** [comment]
@dalefugier the linked issue is cross-platform including Xamarin.Mac, see https://github.com/mono/mono/issues/10304#issuecomment-450736756
I'll check your repro case just in case

**Comment by dalefugier** [comment]
@EgorBo, did you ever have a chance to look at the example I posted?

**Comment by dbelcher** [comment]
@marek-safar When can we expect a fix for this bug to make its way into stable Mono?  It looks like this and #10304 are fixed and in preview versions of Mono, but it would be helpful to know an approximate time period for graduating to stable.  Thanks.

**Comment by marek-safar** [comment]
@dbelcher the fix should be available in stable Mono versions since Mono 5.20

**Comment by dbelcher** [comment]
@marek-safar Thank you for the reply and the help interpreting this one.  We are using the Visual Studio for Mac channel for our embedded Mono, so that explains it.  Looking forward to seeing this one.   Much appreciated.

**Comment by dalefugier** [comment]
@EgorBo, @marek-safar, we just tested the above code with Mono MDK 5.20.1.19, and the issue still exists. Can you re-open this issue?

**Comment by marek-safar** [comment]
@dalefugier are you saying it's not duplicate of the issue mentioned above or both of them are still failing with 5.20 ?

**Comment by dalefugier** [comment]
@marek-safar, apparently this issue (12995) is _not_ related to the linked issue [(10304)](https://github.com/mono/mono/issues/10304), as I can still repeat what I reported (above) with Mono MDK 5.20.1.19. Make sense?

**Comment by dbelcher** [comment]
Hi @marek-safar Is there any reasonable expectation that this could get fixed soon?
This remains an issue for us with Mono 6.4.0.198.   The bug on our end that is a symptom of this bug:
https://mcneel.myjetbrains.com/youtrack/issue/RH-49443
remains a major annoyance for our users.

**Comment by marek-safar** [comment]
@steveisok please allocate some time to estimate fixes for this issue

**Comment by steveisok** [comment]
Appears our faultcode parsing was too 'by the book' as it tried to read the content as an XmlQualifiedName.  This meant that `<faultcode>s:1</faultcode>` was being flagged as invalid because the xml spec says that the NameStartChar cannot be a number.  

I fixed this locally by taking the .net framework block that reads in the fault code.  It's much more forgiving ;-).  Only problem is I need to determine if I broke anything else.
